Sources/AutomergeRepo/Networking/Providers/WebSocketProvider.swift

```diff
@@ -394,10 +394,50 @@ public final class WebSocketProvider: NetworkProvider {
                         )
                 }
                 reconnectAttempts += 1
+                                
                 do {
-                    try await Task.sleep(for: .seconds(waitBeforeReconnect))
-                    // if endpoint is nil, this returns nil
-                    _ = try await attemptConnect(to: endpoint)
+                    // Wait to reconnect. Wait for waitBeforeReconnect and networth path
+                    // transitioning from not satisfied to satisfied. Whichever comes first.
+                    let success = try await withThrowingTaskGroup(of: Void.self) { group in
+
+                        // Wait for timeout
+                        group.addTask {
+                            try await Task.sleep(for: .seconds(waitBeforeReconnect))
+                        }
+                        
+                        // Wait for network becomming availible
+                        group.addTask {
+                            let monitor = NWPathMonitor()
+                            var last = monitor.currentPath
+                            for await each in monitor.paths() {
+                                if last.status != .satisfied && each.status == .satisfied {
+                                    Logger.websocket.info("WEBSOCKET: Network path satisfied while waiting to reconnect")
+                                    return
+                                } else {
+                                    last = each
+                                }
+                            }
+                        }
+                        
+                        // This is what I want, but warns of data races
+                        // _ = try await group.next()
+                        // group.cancelAll()
+                        // return true
+                        
+                        // This accomplishes same thing as previous comment, but doesn't warn of data races
+                        for try await _ in group {
```

Comment:
This works, but it's not doing _quite_ what I think you want - it's waiting for both of the tasks to be complete, in whatever order they happen. I'm not clear enough on if ordering is "first one to complete" or "in the order they're specified above", but which-ever happens results in a retry happening when the first result comes through.

The earlier version is the "I want to intentionally race these two tasks", and loose variation of what I did racing a task that did something against a timeout, with the idea which-ever got done first would provide the result (if any) and the remaining task would be cancelled, and asynchronous execution would continue.

The error you're seeing is about non-isolated access (I ran into it too, earlier) - and the reason is because taskGroup returns a non-isolated closure inside what's currently an isolated function. The way to work around it (and you'll see it in the timeout setup) is to make the function hosting the method `nonisolated`, which resolves the issue. It may also be resolved without that in Swift 6, but I'm guessing and VERY out over my skis on any assertions along those lines. _however_, if we make this private function nonisolated, then the side effects that we're relying on happening **aren't** able to be used - because if the whole function is nonisolated, it doesn't have unfettered access to the internals of the class that we're wanting to update as side effects (peered = true/false, webtask to nil or not, etc).

I **think** we can work around this by isolating this whole "race" of the delay vs. NWPath state changes by breaking that logic out into it's own separate `nonisolated` async functionm and then calling it from within this loop. I'll play some tomorrow to see if I can wrangle that.

The downside of this heuristic, which isn't fatal - but might not be desired - is that if the timeout expires _before_ the NWPath state either is live or switches to live, then it'll try to make a connection even if there's no way it can succeed (NWPath says so) rather than jumping forward without a retry attempt to wait until with a potentially longer delay.

Reply:
It's definitely working, and I'm glad to learn that `taskgroup.next()` returns in order of completion instead of in order of creation - which after I was wondering about, I looked - and [the docs for it](https://developer.apple.com/documentation/swift/taskgroup/next()) do lay it out, I just hadn't gotten there. Too pre-occupied with the pattern that explicitly called next() once (outside of a loop) and all the quirkiness with it due to isolated & nonisolated implications. (Hopefully that'll be rectified in Swift 6)

I'd suggest maybe removing the commented out code with "I wanted to do this..." - that's what sent me down that nonisolated rabbit hole, as the break & return on the first item received in the loop works perfectly in this scenario. This code does still make attempts, even when it knows it'll fail, but it's an improvement over not having this at all in that it'll now short-circuit and check immediately if the triggering change isn't the timeout, but the network coming available.

I'm also thrilled you explored creating NWPathMonitor() én mass, that is definitely easier to pick up when we need it and not deal with continuously flowing state if we don't have to. So delete the bits between line 422 and 425, really so there's no confusion/WIP left in place, and I think this is good to merge!

Reply:
Thanks for taking a look at it all. I've clean up the comments.